albatross.ml

```diff
@@ -0,0 +1,299 @@
+                Vmm_commands.pp_version Vmm_commands.current);
+        Logs.debug (fun m ->
+            m "albatross returned: %a"
+              (Vmm_commands.pp_wire ~verbose:true)
+              (hdr, res));
+        w
+
+  let decode_reply data =
+    if Cstruct.length data >= 4 then
+      let len = Int32.to_int (Cstruct.BE.get_uint32 data 0) in
+      if Cstruct.length data >= 4 + len then (
+        if Cstruct.length data > 4 + len then
+          Logs.warn (fun m ->
+              m "received %d trailing bytes" (Cstruct.length data - len - 4));
+        decode (Cstruct.sub data 4 len))
+      else (
+        Logs.err (fun m ->
+            m "buffer too short (%d bytes), need %d + 4 bytes"
+              (Cstruct.length data) len);
+        Error ())
+    else (
+      Logs.err (fun m ->
+          m "buffer too short (%d bytes), need at least 4 bytes"
+            (Cstruct.length data));
+      Error ())
+
+  let split_many data =
+    let rec split acc data =
+      if Cstruct.length data >= 4 then
+        let len = Int32.to_int (Cstruct.BE.get_uint32 data 0) in
+        if Cstruct.length data >= 4 + len then
+          split (Cstruct.sub data 4 len :: acc) (Cstruct.shift data (len + 4))
+        else (
+          Logs.warn (fun m ->
+              m "buffer too small: %u bytes, requires %u bytes"
+                (Cstruct.length data - 4)
+                len);
+          acc)
+      else if Cstruct.length data = 0 then acc
+      else (
+        Logs.warn (fun m ->
+            m "buffer too small: %u bytes leftover" (Cstruct.length data));
+        acc)
+    in
+    split [] data |> List.rev
+
+  let rec continue_reading t name flow =
+    let open Lwt.Infix in
+    TLS.read flow >>= function
+    | Ok (`Data d) ->
+        let bufs = split_many d in
+        List.iter
+          (fun d ->
+            match decode d with
+            | Error () -> ()
+            | Ok (_, `Data (`Console_data (ts, data))) ->
+                let d = Albatross_json.console_data_to_json (ts, data) in
+                t.console_output <-
+                  Map.update name
+                    (function
+                      | None -> Some [ d ]
+                      | Some xs ->
+                          let xs =
+                            if List.length xs > 20 then List.tl xs else xs
+                          in
+                          Some (xs @ [ d ]))
+                    t.console_output
+            | _ ->
+                Logs.warn (fun m ->
+                    m "unexpected reply, expected console output"))
+          bufs;
+        continue_reading t name flow
+    | Ok `Eof ->
+        t.console_readers <- Set.remove name t.console_readers;
+        TLS.close flow
+    | Error e ->
+        t.console_readers <- Set.remove name t.console_readers;
+        TLS.close flow >|= fun () ->
+        Logs.err (fun m -> m "received error while reading: %a" TLS.pp_error e)
+
+  let query t ?domain ?(name = ".") cmd =
```

Comment:
Some thoughts, 
let's say we have this endpoint `/unikernel/info/domain:name` 
and somehow a user figures out they can edit the link to `/unikernel/info/`, could this function "potentially" return the result which would be the list of all unikernels?

Reply:
We never should have the `domain` exposed in an endpoint. The idea is that the `domain` will always be added by mollymawk depending on the user account. And the client can only choose the unikernel name.

Reply:
At the moment, this is all in flux, since the old code and interface didn't deal with the policies properly. But I'm sure we will adapt the code to always use authentication and then to pass the domain properly to albatross :)

Reply:
What I intended to say: good thinking, this is a very valid question. And indeed we need to be careful to not expose that.

As aside, what we developed in Marrakesh was cutting of the ":" from the unikernel names -- there we didn't bother about domains and the policies. Now, a `/unikernel/info/:robur:roburtls` won't work -- since albatross rejects the `:` in the name (only allows letters, digits, hyphen). In order to access the robur unikernels we need, as done in this PR, the intermediate certificate.

And you are very right, mollymawk has the permissions with the certificate and key it holds to create/destroy _all the unikernels_, so we really need to be careful to not expose functionality meant for other users within the HTTP interface. :)

Reply:
Okay great. I'm happy to know that the `domain` is added later on at some point during processing and the user can never know what their domain is. I think information like this shouldn't be added to any user endpoints, maybe system admins (`Robur`)

Reply:
Now that we're talking about it, maybe the path to go is to make the `domain` argument mandatory :) I went ahead and pushed 7f7128b